services/mailer/mail_issue.go

```diff
@@ -94,78 +99,69 @@ func mailIssueCommentToParticipants(ctx *mailCommentContext, mentions []int64) e
 		visited[i] = true
 	}
 
-	if err = mailIssueCommentBatch(ctx, unfiltered, visited, false); err != nil {
+	unfilteredUsers, err := models.GetMaileableUsersByIDs(unfiltered, false)
+	if err != nil {
+		return err
+	}
+	if err = mailIssueCommentBatch(ctx, unfilteredUsers, visited, false); err != nil {
 		return fmt.Errorf("mailIssueCommentBatch(): %v", err)
 	}
 
 	return nil
 }
 
-func mailIssueCommentBatch(ctx *mailCommentContext, ids []int64, visited map[int64]bool, fromMention bool) error {
-	const batchSize = 100
-	for i := 0; i < len(ids); i += batchSize {
-		var last int
-		if i+batchSize < len(ids) {
-			last = i + batchSize
-		} else {
-			last = len(ids)
-		}
-		unique := make([]int64, 0, last-i)
-		for j := i; j < last; j++ {
-			id := ids[j]
-			if _, ok := visited[id]; !ok {
-				unique = append(unique, id)
-				visited[id] = true
-			}
-		}
-		recipients, err := models.GetMaileableUsersByIDs(unique, fromMention)
-		if err != nil {
-			return err
+func mailIssueCommentBatch(ctx *mailCommentContext, users []*models.User, visited map[int64]bool, fromMention bool) error {
+	checkUnit := models.UnitTypeIssues
+	if ctx.Issue.IsPull {
+		checkUnit = models.UnitTypePullRequests
+	}
+
+	langMap := make(map[string][]string)
+	for _, user := range users {
+		// At this point we exclude:
+		// user that don't have all mails enabled or users only get mail on mention and this is one ...
+		if !(user.EmailNotificationsPreference == models.EmailNotificationsEnabled ||
+			fromMention && user.EmailNotificationsPreference == models.EmailNotificationsOnMention) {
+			continue
 		}
 
-		checkUnit := models.UnitTypeIssues
-		if ctx.Issue.IsPull {
-			checkUnit = models.UnitTypePullRequests
+		// if we have already visited this user we exclude them
+		if _, ok := visited[user.ID]; ok {
+			continue
 		}
-		// Make sure all recipients can still see the issue
-		idx := 0
-		for _, r := range recipients {
-			if ctx.Issue.Repo.CheckUnitUser(r, checkUnit) {
-				recipients[idx] = r
-				idx++
-			}
+
+		// now mark them as visited
+		visited[user.ID] = true
+
+		// test if this user is allowed to see the issue/pull
+		if !ctx.Issue.Repo.CheckUnitUser(user, checkUnit) {
+			continue
 		}
-		recipients = recipients[:idx]
 
-		// TODO: Separate recipients by language for i18n mail templates
-		tos := make([]string, len(recipients))
-		for i := range recipients {
-			tos[i] = recipients[i].Email
+		langMap[user.Language] = append(langMap[user.Language], user.Email)
+	}
+
+	for lang, receivers := range langMap {
+		for i := ((len(receivers) - 1) / MailBatchSize) * MailBatchSize; i >= 0; i -= MailBatchSize {
```

Comment:
```suggestion
		// because we know that the len(receivers) > 0 and we don't care about the order particularly
		// working backwards from the last (possibly) incomplete batch. If len(receivers) can be 0 this
		// starting condition will need to be changed slightly
		for i := ((len(receivers) - 1) / MailBatchSize) * MailBatchSize; i >= 0; i -= MailBatchSize {
```